File: 844.py
# 844. Backspace String Compare
# Given two strings s and t, return true if they are equal when both are typed into empty text editors. '#' means a backspace character.
# Note that after backspacing an empty text, the text will continue empty.

class Solution:
    def backspaceCompare(self, s: str, t: str) -> bool:
        i = len(s) - 1
        typedS = ''
        backspaces = 0
        while i >= 0:
            if s[i] == '#':
                backspaces += 1
            elif backspaces:
                backspaces -= 1
            else:
                typedS += s[i]
            i -= 1
        i = len(t) - 1
        typedT = ''
        backspaces = 0
        while i >= 0:
            if t[i] == '#':
                backspaces += 1
            elif backspaces:
                backspaces -= 1
            else:
                typedT += t[i]
            i -= 1
        return typedS == typedT

s = Solution()
print(s.backspaceCompare('ab#c', 'ad#c'))
print(s.backspaceCompare('ab##', 'c#d#'))
print(s.backspaceCompare('a#c', 'b'))

File: 844.js
// 844. Backspace String Compare
// Given two strings s and t, return true if they are equal when both are typed into empty text editors. '#' means a backspace character.
// Note that after backspacing an empty text, the text will continue empty.

var backspaceCompare = function(s, t) {
    let typedS = ''
    let backspaces = 0
    for (let i = s.length - 1; i >= 0; i--) {
        let char = s[i]
        if (char === '#') {
            backspaces++
        } else if (backspaces) {
            backspaces--
        } else {
            typedS += char
        }
        if (backspaces >= i) break
    }
    let typedT = ''
    backspaces = 0
    for (let i = t.length - 1; i >= 0; i--) {
        let char = t[i]
        if (char === '#') {
            backspaces++
        } else if (backspaces) {
            backspaces--
        } else {
            typedT += char
        }
        if (backspaces >= i) break
    }
    return typedS === typedT
};

console.log(backspaceCompare("ab#c", "ad#c"))
console.log(backspaceCompare("ab##", "c#d#"))
console.log(backspaceCompare("a#c", "b"))